src/checkers/cwe_476.ml

```diff
@@ -1,47 +1,280 @@
+      let () = add_to_hits cwe_hits tid in
+      State.remove_tid state var
+    | NoAccess -> state
+  end in
+  match Jmp.kind jmp with
+  | Goto(Indirect(exp)) -> flag_any_access exp state cwe_hits
+  | Goto(Direct(_)) -> state
+  | Ret(_) -> if strict_call_policy then
+      flag_all_unchecked_registers state cwe_hits
+    else
+      state
+  | Int(_, _) -> flag_all_unchecked_registers state cwe_hits
+  | Call(call) ->
+    let state = match Call.return call with
+      | Some(Indirect(exp)) -> flag_any_access exp state cwe_hits
+      | _ -> state in
+    let state = match Call.target call with
+      | Indirect(exp) -> flag_any_access exp state cwe_hits
+      | _ -> state in
+    let state = match strict_call_policy with
+      | true -> (* all unchecked registers get flagged as hits *)
+        flag_all_unchecked_registers state cwe_hits
+      | false -> (* we assume that the callee will check all remaining unchecked values *)
+        [] in
+    match Call.target call with
+    | Indirect(_) -> state (* already handled above *)
+    | Direct(tid) ->
+      if List.exists function_names ~f:(fun elem -> String.(=) elem (Tid.name tid)) then
+        let func = Term.find_exn sub_t program tid in
+        let arguments = Term.enum arg_t func in
+        (* Every return register is tainted as unchecked return value. *)
+        Seq.fold arguments ~init:state ~f:(fun state arg ->
+            match Bap.Std.Arg.intent arg with
+            | None | Some(In) -> state
+            | Some(Out) | Some(Both) ->
+              let variable = match Bap.Std.Arg.rhs arg with
+                | Bil.Var(var) -> var
+                | _ -> failwith "[CWE476] Return register wasn't a register." in
+              State.add state variable (Term.tid block)
+          )
+      else
+        state
+
+(** updates a block analysis.
+    The strict call policy decides the behaviour on call and return instructions:
+    strict: all unchecked values get flagged as cwe-hits
+    non-strict: the state gets cleared, it is assumed that the target of the call/return
+    instruction checks all remaining unchecked values. *)
+let update_block_analysis block register_state ~cwe_hits ~function_names ~program ~strict_call_policy =
+  let elements = Blk.elts block in
+  let register_state = Seq.fold elements ~init:register_state ~f:(fun state element ->
+      match element with
+      | `Def def -> update_state_def def state ~cwe_hits
+      | `Phi phi -> state (* We ignore phi terms for this analysis. *)
+      | `Jmp jmp -> update_state_jmp jmp state ~cwe_hits ~function_names ~program ~block ~strict_call_policy
+    ) in
+  State.remove_virtual_registers register_state (* virtual registers should not be accessed outside of the block where they are defined. *)
+
+let print_hit tid ~sub ~function_names ~tid_map =
+  let block = Option.value_exn (Term.find blk_t sub tid) in
+  let jmps = Term.enum jmp_t block in
+  let _ = Seq.find_exn jmps ~f:(fun jmp ->
```

Comment:
Seq.find_exn seems the most logical choice for me, because it is supposed to find exactly one element and it throws an exception if it doesn't find anything. Usually I would use the return value for the Log_utils.warn call, but then I would have to unpack the jmp again.